mint_utils.go

```diff
@@ -0,0 +1,150 @@
+package quic
+
+import (
+	"bytes"
+	gocrypto "crypto"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"io"
+
+	"github.com/bifurcation/mint"
+	"github.com/lucas-clemente/quic-go/internal/crypto"
+	"github.com/lucas-clemente/quic-go/internal/handshake"
+	"github.com/lucas-clemente/quic-go/internal/protocol"
+	"github.com/lucas-clemente/quic-go/internal/utils"
+	"github.com/lucas-clemente/quic-go/internal/wire"
+)
+
+type mintController struct {
+	csc  *handshake.CryptoStreamConn
+	conn *mint.Conn
+}
+
+var _ handshake.MintTLS = &mintController{}
+
+func newMintController(
+	csc *handshake.CryptoStreamConn,
+	mconf *mint.Config,
+	pers protocol.Perspective,
+) handshake.MintTLS {
+	var conn *mint.Conn
+	if pers == protocol.PerspectiveClient {
+		conn = mint.Client(csc, mconf)
+	} else {
+		conn = mint.Server(csc, mconf)
+	}
+	return &mintController{
+		csc:  csc,
+		conn: conn,
+	}
+}
+
+func (mc *mintController) GetCipherSuite() mint.CipherSuiteParams {
+	return mc.conn.State().CipherSuite
+}
+
+func (mc *mintController) ComputeExporter(label string, context []byte, keyLength int) ([]byte, error) {
+	return mc.conn.ComputeExporter(label, context, keyLength)
+}
+
+func (mc *mintController) Handshake() mint.Alert {
+	return mc.conn.Handshake()
+}
+
+func (mc *mintController) State() mint.State {
+	return mc.conn.State().HandshakeState
+}
+
+func (mc *mintController) SetCryptoStream(stream io.ReadWriter) {
+	mc.csc.SetStream(stream)
+}
+
+func (mc *mintController) SetExtensionHandler(h mint.AppExtensionHandler) error {
+	return mc.conn.SetExtensionHandler(h)
+}
+
+func tlsToMintConfig(tlsConf *tls.Config, pers protocol.Perspective) (*mint.Config, error) {
+	mconf := &mint.Config{
+		NonBlocking: true,
+		CipherSuites: []mint.CipherSuite{
+			mint.TLS_AES_128_GCM_SHA256,
+			mint.TLS_AES_256_GCM_SHA384,
+		},
+	}
+	if tlsConf != nil {
+		mconf.Certificates = make([]*mint.Certificate, len(tlsConf.Certificates))
+		for i, certChain := range tlsConf.Certificates {
+			mconf.Certificates[i] = &mint.Certificate{
+				Chain:      make([]*x509.Certificate, len(certChain.Certificate)),
+				PrivateKey: certChain.PrivateKey.(gocrypto.Signer),
+			}
+			for j, cert := range certChain.Certificate {
+				c, err := x509.ParseCertificate(cert)
+				if err != nil {
+					return nil, err
+				}
+				mconf.Certificates[i].Chain[j] = c
+			}
+		}
+	}
+	if err := mconf.Init(pers == protocol.PerspectiveClient); err != nil {
+		return nil, err
+	}
+	return mconf, nil
+}
+
+// unpackInitialOrRetryPacket unpacks packets Initial and Retry packets
+// These packets must contain a STREAM_FRAME for the crypto stream, starting at offset 0.
+func unpackInitialPacket(aead crypto.AEAD, hdr *wire.Header, data []byte, version protocol.VersionNumber) (*wire.StreamFrame, error) {
```

Comment:
Where do we check the initial packet length?

Reply:
Not yet, since we're not sending the right length yet: #886 will fix this once this PR is merged.